pymc3/distributions/mixture.py

```diff
@@ -108,86 +130,220 @@ def comp_dists(self):
+        if self.is_multidim_comp:
+            samples = self._comp_dists.random(point=point, size=size)
         else:
-            # We must iterate the calls to random manually
-            size = to_tuple(size)
-            _size = int(np.prod(size))
-            try:
-                samples = np.array([self.comp_dists.random(point=point,
-                                                           size=None)
-                                    for _ in range(_size)])
-                samples = np.reshape(samples, size + samples.shape[1:])
-            except AttributeError:
-                samples = np.array([[comp_dist.random(point=point, size=None)
-                                     for _ in range(_size)]
-                                    for comp_dist in self.comp_dists])
-                samples = np.moveaxis(samples, 0, samples.ndim - 1)
-                samples = np.reshape(samples, size + samples[1:])
-
-        if samples.shape[-1] == 1:
-            return samples[..., 0]
+            if comp_dist_shapes is None:
+                comp_dist_shapes = self._comp_dist_shapes
+            if broadcast_shape is None:
+                broadcast_shape = self._sample_shape
+            samples = []
+            for dist_shape, generator in zip(comp_dist_shapes,
+                                             self._generators):
+                sample = generate_samples(
+                    generator=generator,
+                    dist_shape=dist_shape,
+                    broadcast_shape=broadcast_shape,
+                    point=point,
+                    size=size,
+                    not_broadcast_kwargs={'raw_size_': size},
+                )
+                samples.append(sample)
+            samples = np.array(
+                broadcast_distribution_samples(samples, size=size)
+            )
+            # In the logp we assume the last axis holds the mixture components
+            # so we move the axis to the last dimension
+            samples = np.moveaxis(samples, 0, -1)
+        return samples.astype(self.dtype)
+
+    def infer_comp_dist_shapes(self, point=None):
+        if self.is_multidim_comp:
+            if len(self._comp_dist_shapes) > 0:
+                comp_dist_shapes = self._comp_dist_shapes
+            else:
+                # Happens when the distribution is a scalar or when it was not
+                # given a shape. In these cases we try to draw a single value
+                # to check its shape, we use the provided point dictionary
+                # hoping that it can circumvent the Flat and HalfFlat
+                # undrawable distributions.
+                with _DrawValuesContextBlocker():
+                    test_sample = self._comp_dists.random(point=point,
+                                                          size=None)
+                    comp_dist_shapes = test_sample.shape
+            broadcast_shape = comp_dist_shapes
+            includes_mixture = True
```

Comment:
I am not complete sure what is the function of `includes_mixture` here - I imagine is something to do with making sure the broadcasting of `w` is correct. Could you add some small docstrings to this function?

Reply:
You're right. Yes, I'll add a docstring. `includes_mixture` maybe was not the best choice of name, and maybe is redundant. It's a bool that indicates whether the `comp_dists` last axis has the mixture components or not, but that's the same as testing if `comp_dists` is a distribution or a list. Hmm, I'll see how not to depend on it.